File: experiments-local/experiment-17/runs/agent=omp_language=typescript_model=openrouter/z-ai/glm-5.2_tooling=none/rep1/src/tools.ts
/**
 * Brazilian Soccer MCP Server — Tool registry.
 *
 * Context block
 * -------------
 * Exposes the five required query categories as MCP tools. Each tool is a thin
 * wrapper around an exported, pure handler function (`*Handler`) that takes a
 * typed args object plus the loaded `Dataset` and returns the answer text.
 * Keeping handlers separate from the MCP `server.tool()` registration means
 * the BDD suite can exercise the exact tool logic without spinning up a
 * transport, while `registerTools` wires them to the MCP server for real
 * clients (Claude Desktop, etc.).
 *
 * Tools:
 *   - search_matches        : find matches by team / opponent / competition / season / date range
 *   - team_statistics       : win/loss/draw record, goals, optional venue + competition split
 *   - compare_teams         : head-to-head + parallel record comparison
 *   - search_players        : FIFA player search by name/nationality/club/position/rating
 *   - competition_standings : computed standings, champion, relegation
 *   - match_statistics      : avg goals, home/away win rates, biggest victories
 *   - list_teams            : enumerate known teams (helps LLM disambiguate names)
 *   - list_competitions     : enumerate competitions + seasons present in data
 */

import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { Dataset } from "./data/loader.js";
import {
  biggestWins,
  findMatches,
  findPlayers,
  formatHeadToHead,
  formatMatch,
  formatPlayer,
  formatStandings,
  formatTeamRecord,
  headToHead,
  matchStats,
  resolveTeam,
  standings,
  teamRecord,
} from "./data/query.js";
import type { Competition } from "./data/types.js";

const COMPETITIONS: Competition[] = [
  "Brasileirão",
  "Copa do Brasil",
  "Copa Libertadores",
  "Historical Brasileirão",
  "Other",
];

/** Shared zod enum for competition filters across tools. */
const competitionEnum = z.enum([
  "Brasileirão",
  "Copa do Brasil",
  "Copa Libertadores",
  "Historical Brasileirão",
  "Other",
  "all",
]);

function asComp(v: string | undefined): Competition | "all" {
  return (v as Competition | "all") ?? "all";
}

/** Wrap a plain-text answer as an MCP CallToolResult payload. */
function textResult(text: string): { content: [{ type: "text"; text: string }] } {
  return { content: [{ type: "text", text }] };
}

// ---- search_matches ---------------------------------------------------------

export interface SearchMatchesArgs {
  team?: string;
  opponent?: string;
  competition?: string;
  season?: number;
  start_date?: string;
  end_date?: string;
  limit?: number;
}

export function searchMatchesHandler(ds: Dataset, args: SearchMatchesArgs): string {
  const results = findMatches(ds, {
    team: args.team,
    opponent: args.opponent,
    competition: asComp(args.competition),
    season: args.season,
    startDate: args.start_date,
    endDate: args.end_date,
    limit: args.limit ?? 50,
  });
  if (results.length === 0) return "No matches found for the given criteria.";
  return `Found ${results.length} match(es):\n` + results.map(formatMatch).join("\n");
}

// ---- team_statistics --------------------------------------------------------

export interface TeamStatisticsArgs {
  team: string;
  competition?: string;
  season?: number;
  venue?: "home" | "away" | "any";
}

export function teamStatisticsHandler(ds: Dataset, args: TeamStatisticsArgs): string {
  const matches = findMatches(ds, {
    team: args.team,
    competition: asComp(args.competition),
    season: args.season,
  });
  const resolved = resolveTeam(ds, args.team) ?? args.team;
  const rec = teamRecord(matches, resolved, args.venue === "any" || !args.venue ? undefined : args.venue);
  const labelParts = [
    resolved,
    args.season ? String(args.season) : undefined,
    args.competition && args.competition !== "all" ? args.competition : undefined,
    args.venue && args.venue !== "any" ? `${args.venue} record` : "record",
  ];
  return formatTeamRecord(rec, labelParts.filter(Boolean).join(" — "));
}

// ---- compare_teams ----------------------------------------------------------

export interface CompareTeamsArgs {
  team_a: string;
  team_b: string;
  competition?: string;
}

export function compareTeamsHandler(ds: Dataset, args: CompareTeamsArgs): string {
  const h2h = headToHead(ds, args.team_a, args.team_b);
  const comp = asComp(args.competition);
  const recA = teamRecord(findMatches(ds, { team: args.team_a, competition: comp }), h2h.teamA);
  const recB = teamRecord(findMatches(ds, { team: args.team_b, competition: comp }), h2h.teamB);
  return [
    formatHeadToHead(h2h),
    "",
    formatTeamRecord(recA, `${h2h.teamA} overall`),
    "",
    formatTeamRecord(recB, `${h2h.teamB} overall`),
  ].join("\n");
}

// ---- search_players ---------------------------------------------------------

export interface SearchPlayersArgs {
  name?: string;
  nationality?: string;
  club?: string;
  position?: string;
  min_overall?: number;
  sort_by?: "overall" | "potential" | "age" | "name";
  limit?: number;
}

export function searchPlayersHandler(ds: Dataset, args: SearchPlayersArgs): string {
  const results = findPlayers(ds, {
    name: args.name,
    nationality: args.nationality,
    club: args.club,
    position: args.position,
    minOverall: args.min_overall,
    sortBy: args.sort_by,
    limit: args.limit ?? 25,
  });
  if (results.length === 0) return "No players found matching the criteria.";
  return `Found ${results.length} player(s):\n` + results.map(formatPlayer).join("\n");
}

// ---- competition_standings --------------------------------------------------

export interface CompetitionStandingsArgs {
  competition: string;
  season: number;
  limit?: number;
}

export function competitionStandingsHandler(ds: Dataset, args: CompetitionStandingsArgs): string {
  const full = standings(ds, asComp(args.competition), args.season);
  if (full.length === 0) return `No standings data for ${args.competition} ${args.season}.`;
  const rows = full.slice(0, args.limit ?? 30);
  return formatStandings(rows, `${args.season} ${args.competition} Standings (computed)`, full.length);
}

// ---- match_statistics -------------------------------------------------------

export interface MatchStatisticsArgs {
  competition?: string;
  season?: number;
  biggest_wins?: number;
}

export function matchStatisticsHandler(ds: Dataset, args: MatchStatisticsArgs): string {
  const matches = findMatches(ds, {
    competition: asComp(args.competition),
    season: args.season,
  });
  const stats = matchStats(matches);
  const pct = (x: number) => (x * 100).toFixed(1) + "%";
  const lines = [
    `Statistics for ${args.competition ?? "all"}${args.season ? " " + args.season : ""}:`,
    `- Matches with scores: ${stats.matches}`,
    `- Total goals: ${stats.totalGoals}`,
    `- Average goals per match: ${stats.avgGoalsPerMatch.toFixed(2)}`,
    `- Home win rate: ${pct(stats.homeWinRate)}`,
    `- Draw rate: ${pct(stats.drawRate)}`,
    `- Away win rate: ${pct(stats.awayWinRate)}`,
  ];
  const n = args.biggest_wins ?? 5;
  if (n > 0) {
    const wins = biggestWins(matches, n);
    if (wins.length > 0) {
      lines.push("", "Biggest victories:");
      wins.forEach((m, i) => lines.push(`${i + 1}. ${formatMatch(m).slice(2)}`));
    }
  }
  return lines.join("\n");
}

// ---- list_teams -------------------------------------------------------------

export interface ListTeamsArgs {
  query?: string;
  limit?: number;
}

export function listTeamsHandler(ds: Dataset, args: ListTeamsArgs): string {
  let teams = ds.teams.all().sort((a, b) => a.localeCompare(b));
  if (args.query) {
    const q = args.query.toLowerCase();
    teams = teams.filter((t) => t.toLowerCase().includes(q));
  }
  teams = teams.slice(0, args.limit ?? 100);
  return `Known teams (${teams.length} shown):\n${teams.join("\n")}`;
}

// ---- list_competitions ------------------------------------------------------

export function listCompetitionsHandler(ds: Dataset): string {
  const lines = ["Competitions in dataset:"];
  for (const c of COMPETITIONS) {
    const seasons = (ds.seasonsByCompetition.get(c) ?? []).sort((a, b) => a - b);
    const count = ds.matchesByCompetition.get(c)?.length ?? 0;
    lines.push(`- ${c}: ${count} matches, seasons: ${seasons.length ? seasons.join(", ") : "n/a"}`);
  }
  return lines.join("\n");
}

// ---- registration -----------------------------------------------------------

/** Register all Brazilian-soccer tools on the given server. */
export function registerTools(server: McpServer, ds: Dataset): void {
  server.tool(
    "search_matches",
    "Find Brazilian soccer matches by team, opponent, competition, season, and/or date range. " +
      "Returns a formatted list of matches with date, score, and competition context.",
    {
      team: z.string().optional().describe("Team name (home, away, or either). Accepts variants like 'Palmeiras-SP'."),
      opponent: z.string().optional().describe("Opponent team — when set with `team`, returns only matches between the two."),
      competition: competitionEnum.optional().describe("Competition filter; 'all' searches every source."),
      season: z.number().int().optional().describe("Season year, e.g. 2023."),
      start_date: z.string().optional().describe("ISO date lower bound, e.g. '2023-01-01'."),
      end_date: z.string().optional().describe("ISO date upper bound, e.g. '2023-12-31'."),
      limit: z.number().int().min(1).max(500).default(50).describe("Max matches to return."),
    },
    (args) => textResult(searchMatchesHandler(ds, args)),
  );

  server.tool(
    "team_statistics",
    "Compute win/loss/draw record, goals for/against, and points for a team, optionally filtered by competition, season, and venue (home/away).",
    {
      team: z.string().describe("Team name (variants accepted, e.g. 'Corinthians-SP')."),
      competition: competitionEnum.optional(),
      season: z.number().int().optional(),
      venue: z.enum(["home", "away", "any"]).optional().describe("Restrict to home or away matches."),
    },
    (args) => textResult(teamStatisticsHandler(ds, args)),
  );

  server.tool(
    "compare_teams",
    "Compare two teams head-to-head: number of meetings, wins each, draws, and goals. Also returns each team's overall record.",
    {
      team_a: z.string(),
      team_b: z.string(),
      competition: competitionEnum.optional(),
    },
    (args) => textResult(compareTeamsHandler(ds, args)),
  );

  server.tool(
    "search_players",
    "Search the FIFA player database by name, nationality, club, position, and/or minimum overall rating. Returns players sorted by overall rating by default.",
    {
      name: z.string().optional(),
      nationality: z.string().optional().describe("e.g. 'Brazil', 'Argentina'."),
      club: z.string().optional().describe("Club name substring, e.g. 'Flamengo'."),
      position: z.string().optional().describe("Position code, e.g. 'ST', 'LW', 'GK', 'CDM'."),
      min_overall: z.number().int().min(0).max(99).optional(),
      sort_by: z.enum(["overall", "potential", "age", "name"]).optional(),
      limit: z.number().int().min(1).max(200).default(25),
    },
    (args) => textResult(searchPlayersHandler(ds, args)),
  );

  server.tool(
    "competition_standings",
    "Compute the standings table for a competition season from match results (3 pts/win, 1/draw). Returns ranked table; first row is champion, bottom four are in the relegation zone.",
    {
      competition: competitionEnum.describe("Competition. Use 'Brasileirão' for Série A."),
      season: z.number().int().describe("Season year, e.g. 2019."),
      limit: z.number().int().min(1).max(100).default(30),
    },
    (args) => textResult(competitionStandingsHandler(ds, args)),
  );

  server.tool(
    "match_statistics",
    "Aggregate statistics over a match set: average goals per match, home/draw/away win rates, total goals, and the biggest victories. Filter by competition and/or season.",
    {
      competition: competitionEnum.optional(),
      season: z.number().int().optional(),
      biggest_wins: z.number().int().min(0).max(50).default(5).describe("How many biggest victories to list."),
    },
    (args) => textResult(matchStatisticsHandler(ds, args)),
  );

  server.tool(
    "list_teams",
    "List known teams in the dataset, optionally filtered by a name substring. Use to resolve team name variants before querying.",
    {
      query: z.string().optional().describe("Substring filter (case-insensitive)."),
      limit: z.number().int().min(1).max(1000).default(100),
    },
    (args) => textResult(listTeamsHandler(ds, args)),
  );

  server.tool(
    "list_competitions",
    "List the competitions present in the dataset and the seasons available for each.",
    {},
    () => textResult(listCompetitionsHandler(ds)),
  );
}

File: experiments-local/experiment-mu-glm53-brazil/runs/agent=opencode_language=python_model=openrouter/z-ai/glm-5.3-flash_tooling=none/rep3/brazilian_soccer/tools.py
"""MCP tool layer: natural-language router + tool implementations.

The MCP server exposes structured tools (see ``server.py``); an LLM client
is expected to choose the right tool and parameters.  ``answer_question``
additionally provides a deterministic best-effort router so simple
questions can be answered without an LLM: it detects years, team names,
competition words and intent keywords, and dispatches to the matching
``SoccerStore`` method.

Router intent order (first match wins):
1. head-to-head   - two team names + "vs"/"against"/"x"
2. standings      - "standings"/"table"/"champion"/"won"/"relegated" + year
3. derbies        - "derby"/"rivalry"/"clássico"
4. players        - "player"/"who is"/"rating"/"squad"
5. statistics     - "average goals"/"biggest win"/"best home"/"best away"
6. team stats     - "record"/"wins" + one team
7. matches        - fallback: match search with detected teams/season
"""

from __future__ import annotations

import re

from .normalize import strip_accents
from .store import NotFound, SoccerStore

_YEAR_RE = re.compile(r"\b(?:19|20)\d{2}\b")


def _detect_years(question: str) -> list[int]:
    return [int(y) for y in _YEAR_RE.findall(question)]


def _match_forms(display: str) -> list[str]:
    """Text forms a team display name can match in a question.

    "Flamengo-RJ" matches "Flamengo" too; "Atlético-MG" matches
    "Atlético" (popularity disambiguates collisions).
    """
    forms = {strip_accents(display).lower()}
    base = re.split(r"\s*[-–]\s*", display)[0]
    base = strip_accents(base).lower()
    if len(base) >= 4:
        forms.add(base)
    return sorted(forms, key=len, reverse=True)


def _detect_teams(store: SoccerStore, question: str) -> list[str]:
    """Find team mentions by scanning display names (longest first)."""
    q = strip_accents(question).lower()
    found: list[tuple[int, str]] = []
    for key, display in store.registry.display_names().items():
        for form in _match_forms(display):
            idx = q.find(form)
            if idx >= 0:
                found.append((idx, display))
                break
    # Longest display name at each position wins; keep reading order.
    found.sort(key=lambda t: (t[0], -len(t[1])))
    picked: list[str] = []
    last_end = -1
    for idx, display in found:
        if idx >= last_end:
            picked.append(display)
            last_end = idx + len(display)
        if len(picked) == 2:
            break
    return picked


def _detect_competition(store: SoccerStore, question: str) -> str | None:
    q = strip_accents(question).lower()
    if "libertadores" in q:
        return "Copa Libertadores"
    if "copa do brasil" in q or "brazilian cup" in q:
        return "Copa do Brasil"
    if "serie b" in q:
        return "Brasileirão Serie B"
    if "serie c" in q:
        return "Brasileirão Serie C"
    if "serie a" in q or "brasileirao" in q or "brasileirão" in q:
        return "Brasileirão Serie A"
    return None


def answer_question(store: SoccerStore, question: str) -> dict:
    """Best-effort deterministic routing of a natural-language question."""
    q = strip_accents(question).lower()
    years = _detect_years(question)
    year = years[-1] if years else None
    competition = _detect_competition(store, question)
    teams = _detect_teams(store, question)

    # 1. head-to-head
    if len(teams) == 2 and any(w in q for w in (" vs ", " x ", " against", " head-to-head", " head to head")):
        return {"tool": "get_head_to_head",
                "result": store.head_to_head(teams[0], teams[1], competition)}

    # 2. standings / champion / relegation
    if any(w in q for w in ("standings", "table", "champion", " who won",
                            "relegat", "won the")):
        comp = competition or "Brasileirão Serie A"
        if year is None:
            raise NotFound("Which season (year)? e.g. 'Who won the 2019 Brasileirão?'")
        result = store.standings(comp, year, top=6)
        if "relegat" in q:
            result = {"relegation_zone": result.get("relegation_zone")}
        elif any(w in q for w in ("who won", "champion")):
            result = {"champion": result["champion"]}
        return {"tool": "get_standings", "result": result}

    # 3. derbies
    if any(w in q for w in ("derby", "derbies", "rivalry", "classico", "clássico")):
        return {"tool": "get_derbies",
                "result": store.derbies(season=year, competition=competition)}

    # 3b. season comparison ("compare the 2018 and 2019 seasons")
    if "compare" in q and len(years) >= 2:
        comp = competition or "Brasileirão Serie A"
        return {"tool": "compare_seasons",
                "result": store.compare_seasons(comp, years[0], years[1])}

    # 3c. club participation ("what competitions has X played in?")
    if "competition" in q and len(teams) == 1:
        return {"tool": "get_team_history",
                "result": store.team_history(teams[0])}

    # 4. players
    if any(w in q for w in ("player", "who is", "rating", "overall", "squad", "forwards")):
        if len(teams) == 1:
            return {"tool": "search_players_at_club",
                    "result": store.players_at_club(teams[0])}
        m = re.search(r"who\s+is\s+(.+?)[\?\.!]?$", question.strip(), re.I)
        if m:
            try:
                return {"tool": "get_player",
                        "result": store.get_player(m.group(1).strip())}
            except NotFound as exc:
                return {"tool": "get_player",
                        "result": {"found": False, "message": str(exc)}}
        if "brazil" in q:
            return {"tool": "search_players",
                    "result": store.search_players(nationality="brazil",
                                                   min_overall=75, limit=10)}
        return {"tool": "search_players",
                "result": store.search_players(name=question, limit=5)}

    # 5. statistics
    if any(w in q for w in ("average goals", "goals per match", "biggest win",
                            "best home", "best away", "home win")):
        return {"tool": "get_statistics",
                "result": store.statistics(competition=competition, season=year)}

    # 6. team stats
    if len(teams) == 1 and any(w in q for w in ("record", "wins", "won how", "how many")):
        venue = None
        if "home" in q:
            venue = "home"
        elif "away" in q:
            venue = "away"
        return {"tool": "get_team_stats",
                "result": store.team_stats(teams[0], season=year,
                                           competition=competition, venue=venue)}

    # 7. match search fallback
    if teams:
        kwargs = {"team": teams[0]}
        if len(teams) == 2:
            kwargs["opponent"] = teams[1]
        if year:
            kwargs["season"] = year
        if competition:
            kwargs["competition"] = competition
        if "last" in q or "when did" in q:
            kwargs["limit"] = 5
            kwargs["order"] = "desc"
        return {"tool": "search_matches", "result": store.search_matches(**kwargs)}

    raise NotFound(
        "Could not route the question. Structured tools are available for "
        "match search, head-to-head, team stats, standings, players, "
        "statistics and derbies.")
